pkg/readiness/object_tracker.go

```diff
@@ -0,0 +1,258 @@
+)
+
+// Expectations tracks expectations for runtime.Objects.
+// A set of Expect() calls are made, demarcated by ExpectationsDone().
+// Expectations are satisfied by calls to Observe().
+// Once all expectations are satisfied, Satisfied() will begin returning true.
+type Expectations interface {
+	Expect(o runtime.Object)
+	CancelExpect(o runtime.Object)
+	ExpectationsDone()
+	Unsatisfied() []runtime.Object
+	Observe(o runtime.Object)
+	Satisfied() bool
+}
+
+// objectTracker tracks expectations for runtime.Objects.
+// A set of Expect() calls are made, demarcated by ExpectationsDone().
+// Expectations are satisfied by calls to Observe().
+// Once all expectations are satisfied, Satisfied() will begin returning true.
+type objectTracker struct {
+	mu        sync.Mutex
+	gvk       schema.GroupVersionKind
+	cancelled objSet // expectations that have been cancelled
+	expect    objSet // unresolved expectations
+	seen      objSet // observations made before their expectations
+	satisfied objSet // tracked to avoid re-adding satisfied expectations and to support Unsatisfied()
+	populated bool   // all expectations have been provided
+}
+
+func newObjTracker(gvk schema.GroupVersionKind) *objectTracker {
+	return &objectTracker{
+		gvk:       gvk,
+		cancelled: make(objSet),
+		expect:    make(objSet),
+		seen:      make(objSet),
+		satisfied: make(objSet),
+	}
+}
+
+// Expect sets an expectation that must be met by a corresponding call to Observe().
+func (t *objectTracker) Expect(o runtime.Object) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Only accept expectations until we're marked as fully populated.
+	if t.populated {
+		return
+	}
+
+	k, err := objKeyFromObject(o)
+	if err != nil {
+		log.Error(err, "skipping")
+		return
+	}
+
+	// Cancelled objects cannot be expected again.
+	if _, ok := t.cancelled[k]; ok {
+		return
+	}
+
+	// We may have seen it before starting to expect it
+	if _, ok := t.seen[k]; ok {
+		delete(t.seen, k)
+		delete(t.expect, k)
+		t.satisfied[k] = struct{}{}
+		return
+	}
+
+	t.expect[k] = struct{}{}
+}
+
+// CancelExpect cancels an expectation and marks it so it
+// cannot be expected again going forward.
+func (t *objectTracker) CancelExpect(o runtime.Object) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	k, err := objKeyFromObject(o)
+	if err != nil {
+		log.Error(err, "skipping")
+		return
+	}
+
+	delete(t.expect, k)
+	delete(t.seen, k)
+	delete(t.satisfied, k)
+	t.cancelled[k] = struct{}{}
+}
+
+// ExpectationsDone tells the tracker to stop accepting new expectations.
+// Only expectations set before ExpectationsDone is called will be considered
+// in Satisfied().
+func (t *objectTracker) ExpectationsDone() {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	log.Info("ExpectationsDone", "gvk", t.gvk, "expectationCount", len(t.expect)+len(t.satisfied))
+	t.populated = true
+}
+
+// Unsatisfied returns all unsatisfied expectations
+func (t *objectTracker) Unsatisfied() []runtime.Object {
+	t.mu.Lock()
```

Comment:
bump

Reply:
Under normal operation I don't expect a lot of contention - each tracker maps to a single controller for writing, plus the readiness probe for reading which won't run that often _under normal circumstances_. I guess since that's an externally facing API it could affect the controllers if someone was slamming `/readyz` in a tight loop. What do you say, still worth going with RWLock?

Reply:
IMO we could get away with it, but RLock is strictly better. Not a blocker for me.

Reply:
Implemented.